### PR TITLE
Add valid usage for resolveQuerySet

### PR DESCRIPTION
**Pull request by haoxli**
Improve the validation rules for resolveQuerySet.

***
<a href="https://pr-preview.s3.amazonaws.com/haoxli/gpuweb/pull/1042.html" title="Last updated on Sep 14, 2020, 3:10 AM UTC (348be63)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/1042/5a4c0f7...haoxli:348be63.html" title="Last updated on Sep 14, 2020, 3:10 AM UTC (348be63)">Diff</a>

### CODE REVIEW

File: spec/index.bs
Comment (Kangz) [inline spec/index.bs:4737]:
```suggestion
                - |destinationOffset| + 8 * |queryCount| is less than or equal to |destination|.{{GPUBuffer/[[size]]}}.
```

File: spec/index.bs
Comment (Kangz) [inline spec/index.bs:4734]: I'm not sure about this one, an alternative is saying that all queries start at 0, so resolving unused queries writes 0 in the buffer.

**Review by kvark — COMMENTED** [review]
LGTM when Corentin's notes are addressed

File: spec/index.bs
Comment (kainino0x) [inline spec/index.bs:4737]: These two can be written as a single rule, as these are Real numbers that can't overflow.
```suggestion
                - |destinationOffset| + 8 &times; |queryCount| &le; |destination|.{{GPUBuffer/[[size]]}}.
```

**Review by kainino0x — APPROVED** [review]
LGTM with resolution on whether queries can be unwritten.

File: spec/index.bs
Comment (haoxli) [reply spec/index.bs:4734]: Vulkan validation layers will throw error if the queries for resolving contain a unused query. So if resolving unused queries to 0 in the buffer, we may need to sumulate by calculating the range of the used queries, then calling resolve method for many times if users skip some indexes when writing values.
For example:
```
encoder.writeTimestamp(querySet, 0);
encoder.writeTimestamp(querySet, 2);
encoder.writeTimestamp(querySet, 3);
encoder.writeTimestamp(querySet, 5);
```

In Vulkan implementaion, we need call vkCmdCopyQueryPoolResults three times:
```
vkCmdCopyQueryPoolResults(cmd, querySet, 0 /*firstIndex*/, 1 /*queryCount*/, dstBuffer, ....);
vkCmdCopyQueryPoolResults(cmd, querySet, 2 /*firstIndex*/, 2 /*queryCount*/, dstBuffer, ....);
vkCmdCopyQueryPoolResults(cmd, querySet, 5 /*firstIndex*/, 1 /*queryCount*/, dstBuffer, ....);
```
The buffer values for other query indexes, we can initialize them to 0, but cannot be worte via vkCmdCopyQueryPoolResults.

Comment (haoxli) [reply spec/index.bs:4734]: Open it as a question in #1072 for more discussion, I will remove this one to land other validations.

**Review by Kangz — APPROVED** [review]

**Review by kainino0x — APPROVED** [review]

**Comment by kvark** [comment]
Looks like this is blocked on #1072 ?

**Comment by haoxli** [comment]
> The queries between [|firstQuery|, |firstQuery| + |queryCount| - 1] are written by query operations.

Only this validation is related to #1072, I have removed it from this PR, so it's not a block now. Wait until we have a conclusion in #1072 before deciding whether to add it back.

**Review by kvark — APPROVED** [review]

**Pull request merged**
